File: mergeSort/mergeSort.go
package main

import (
	"math"
)

// Merges two slices into one ordered slice.
func merge(sliceA, sliceB []int) []int {
	// Contains the elements of both slices in sorted order
	sorted := []int{}

	// Set pointer for both slices at the first elements
	i, j := 0, 0

	// Compare elements of both slices one by one and push
	// them into the sorted slice in ascending order until
	// both pointer reached the end of the respective slice.
	// Every time an element was pushed to the sorted slice,
	// the corresponding pointer moves to the next index.
	for i < len(sliceA) && j < len(sliceB) {
		if sliceA[i] < sliceB[j] {
			sorted = append(sorted, sliceA[i])
			i++
		} else {
			sorted = append(sorted, sliceB[j])
			j++
		}
	}

	// Add the remaining elements from sliceA and sliceB
	// in case they contained an odd number of elements.
	sorted = append(sorted, sliceA[i:]...)
	sorted = append(sorted, sliceB[j:]...)

	// Return a slice with the sorted elements
	return sorted
}

// MergeSort sorts a slice by recursively splitting it
// into subsets of elements until each subset only
// contains one element. It then puts them back
// together in ascending order.
func MergeSort(slice []int) []int {
	size := len(slice)

	// Return the slice if it can't be divided any further
	if size <= 1 {
		return slice
	}

	// Find the index of the median
	middle := int(math.Floor(float64(size) / 2))

	// Recursively sort the left side
	left := MergeSort(slice[0:middle])

	// Recursively sort the right side
	right := MergeSort(slice[middle:])

	// Merge the two sides together in ascending order
	return merge(left, right)
}

File: mergeSort/mergeSort.js
/**
 * Merges two arrays into one ordered array.
 * @param arrayA {array} array of numbers
 * @param arrayB {array} array of numbers
 * @returns {array} sorted array
 */
const merge = (arrayA, arrayB) => {
    // Contains the elements of both arrays in sorted order
    const sorted = [];
    
    // Set pointers for both arrays at the first elements
    let i = 0, j = 0;
    
    // Compare elements of both arrays one by one and push
    // them into the sorted array in ascending order until
    // both pointer reached the end of the respective array.
    // Every time an element was pushed to the sorted array,
    // the corresponding pointer moves to the next index.
    while (i < arrayA.length && j < arrayB.length) {
        if (arrayA[i] < arrayB[j]) {
            sorted.push(arrayA[i]);
            i++;
        } else {
            sorted.push(arrayB[j]);
            j++;
        }
    }
    
    // Return an array with the sorted elements. Add the remaining elements
    // from arrayA and arrayB in case they contained an odd number of elements.
    return [...sorted, ...arrayA.slice(i), ...arrayB.slice(j)];
};

/**
 * Sorts an array by recursively splitting it into
 * subsets of elements until each subset only
 * contains one element. It then puts them back
 * together in ascending order.
 * @param array {array} unsorted array of numbers
 */
const mergeSort = (array) => {
    // Return the array if it can't be divided any further
    if (array.length <= 1) {
        return array;
    }
    
    // Find the index of the median
    const middle = ~~(array.length / 2);
    
    // Recursively sort the left side
    const left = mergeSort(array.slice(0, middle));
    
    // Recursively sort the right side
    const right = mergeSort(array.slice(middle));
    
    // Merge the two sides together in ascending order
    return merge(left, right);
};
